packages/db/src/resources/projects/resolveContractInstances.ts

```diff
@@ -0,0 +1,214 @@
+ * for that past revision.
+ */
+export async function resolveContractInstances(
+  project: IdObject<"projects">,
+  inputs: {
+    contract?: DataModel.ResourceNameInput;
+    network?: DataModel.ResourceNameInput;
+  },
+  context: {
+    workspace: Workspace;
+  },
+  info: graphql.GraphQLResolveInfo
+): Promise<SavedInput<"contractInstances">[]> {
+  const { workspace } = context;
+
+  const contractNameRecords = await resolveNameRecords(
+    project,
+    { ...inputs.contract, type: "Contract" },
+    { workspace }
+  );
+
+  const contractInstances: SavedInput<"contractInstances">[] = [];
+  debug("inputs %O", inputs);
+
+  for await (const { skip, contracts } of findResourcesHistories<"contracts">({
+    collectionName: "contracts",
+    nameRecords: contractNameRecords,
+    workspace
+  })) {
+    let stepContractInstances = await workspace.find("contractInstances", {
+      selector: {
+        "contract.id": { $in: contracts.map(({ id }) => id) }
+      }
+    });
+
+    if (inputs.network) {
+      const ancestors = await filterProjectNetworkAncestors({
+        project,
+        network: inputs.network,
+        candidates: stepContractInstances.map(({ network }) => network),
+        workspace,
+        info
+      });
+
+      const ancestorIds = new Set([...ancestors.map(({ id }) => id)]);
+      stepContractInstances = stepContractInstances.filter(({ network }) =>
+        ancestorIds.has(network.id)
+      );
+    }
+
+    const byContractId = stepContractInstances.reduce(
+      (byContractId, contractInstance) => ({
+        ...byContractId,
+        [contractInstance.contract.id]: contractInstance
+      }),
+      {}
+    );
+
+    const found = contracts.map(({ id }, index) =>
+      id in byContractId ? index : undefined
+    );
+
+    debug("skipping found indexes: %O", found);
+    skip(...found);
+
+    contractInstances.push(...stepContractInstances);
+  }
+
+  return contractInstances;
+}
+
+/**
+ * Steps backwards through name record history for a given set of name records,
+ * yielding an array of past resources in a breadth-first search manner.
+ *
+ * At each step, yields a [somewhat HACKy] `skip` function, to omit specific
+ * indexes from further consideration.
+ *
+ * Returns when all name records histories have been exhausted.
+ */
+async function* findResourcesHistories<N extends NamedCollectionName>(options: {
+  collectionName: N;
+  nameRecords: (SavedInput<"nameRecords"> | undefined)[];
+  workspace: Workspace;
+}): AsyncIterable<
+  {
+    [K in "skip" | N]: "skip" extends K
+      ? (...indexes: number[]) => void
+      : (IdObject<N> | undefined)[];
+  }
+> {
+  const { collectionName, workspace } = options;
+  let { nameRecords } = options;
+
+  do {
+    const skip = (...indexes: (number | undefined)[]) => {
+      for (const index of indexes) {
+        if (typeof index === "number") {
+          nameRecords[index] = undefined;
+        }
+      }
+    };
+
+    // @ts-ignore
```

Comment:
We're racking up `@ts-ignore`'s, haha. Slightly worried about technical debt.

Reply:
Heh this one was just laziness in a place where there's low risk. I can comment to that effect

Reply:
Turned out to be easy enough to be able to remove this.